src/crimson/os/seastore/onode_manager/staged-fltree/node.cc

```diff
@@ -2226,14 +2241,33 @@ Ref<tree_cursor_t> LeafNode::track_insert(
       this, insert_pos);
 }
 
+void LeafNode::reset_values_on_split(
+    context_t c, const search_position_t &split_pos)
+{
+  LOG_PREFIX(LeafNode::reset_values_on_split);
+  DEBUGT("split_pos {}", c.t, split_pos);
+  for (auto iter = tracked_cursors.lower_bound(split_pos);
+      iter != tracked_cursors.end();
+      iter++) {
+    auto p_cursor = iter->second;
+    DEBUGT("resetting {}", c.t, p_cursor->get_position());
+    if (p_cursor->is_mutate()) {
+      p_cursor->reset_value(c);
```

Comment:
Hmm, looks the issue is somehow associated with leaf node split and the modified onodes.

I still need to understand why splitting out a leaf node (i.e. the operation `node_delta_op_t::SPLIT` to the left, right?), the left side (to-trim) can produce inconsistent CRC between runtime and replay-time, when the onode value has been modified.

Does it mean that the leaf node CRC before the operation `node_delta_op_t::SPLIT` was already inconsistent between runtime and replay-time? If so, why?

Reply:
At the present, modifications of onodes' value and the population of onode extents' value recorders are seperated. So the following modification sequence may be problematic:

1. insert onode `Onode X`
2. insert onode `Onode Y` which triggers a split and `Onode X` is moved to the right node.

Splitting a node means copying onodes to the right node and truncating the original node, leaving the fields for those moved onodes intact. However, since the latter insertion which triggers the split happens before populating the value recorders, the modification to the value of `Onode X` is not recorded in the delta of the splitted node, so after system reboot, the replay procedure won't modify the value of `Onode X`, which can lead to crc inconsistency.

Reply:
Sorry for the late reply, and thanks! This root cause is reasonable to me.

> modifications of onodes' value and the population of onode extents' value recorders are seperated

`TestValue` should be immune to this issue, because its modification and delta population are integrated, which was also the intention when the value delta subsystem was initially designed.

That is to say, the same issue is only reproducible in `test_fltree_onode_manager.cc`, not in `test_staged_fltree.cc`, right?

Reply:
If that's true, I think it's not the issue of the low-level staged-tree implementation, it's `FLTreeOnode` not following the supposing invariant.

How about the following fixes?
1. Fix `FLTreeOnode` to populate the delta upon each modification for the correctness (although not efficient yet).
2. Make `FLTreeOnode::Recorder` to manage finer-grained types of deltas just like `TestValue`, instead of the current coarse payload-level delta.

Which also means that the workaround for `FLTreeOnode` in the low-level staged-tree implementations are unnecessary.

Is that reasonable?

Reply:
Yeah, will do.

Reply:
> 1. Fix `FLTreeOnode` to populate the delta upon each modification for the correctness (although not efficient yet).

Hi, yingxin, the current way of modifying `Onode` is first getting a mutable layout of the `Onode` and do any modification to it openly, so we might need to encapsulate each kind of modifications to `Onode` as an method of `onode_layout_t` to `populate the delta upon each modification`, maybe like this:

```
class onode_layout_t {
  // The expected decode size of object_info_t without oid.
  static constexpr int MAX_OI_LENGTH = 232;
  // We might want to move the ss field out of onode_layout_t.
  // The reason is that ss_attr may grow to relative large, as
  // its clone_overlap may grow to a large size, if applications
  // set objects to a relative large size(for the purpose of reducing
  // the number of objects per OSD, so that all objects' metadata
  // can be cached in memory) and do many modifications between
  // snapshots.
  // TODO: implement flexible-sized onode value to store inline ss_attr
  // effectively.
  static constexpr int MAX_SS_LENGTH = 1;
public:
  void update_omap_root(omap_root_t&);
  void update_xattr_root(omap_root_t&);
  void update_object_data(object_data_t&);
  void update_object_info(ceph::bufferlist&);
  void update_snapset(ceph::bufferlist&);

private:
  ceph_le32 size{0};
  ceph_le32 oi_size{0};
  ceph_le32 ss_size{0};
  omap_root_le_t omap_root;
  omap_root_le_t xattr_root;

  object_data_le_t object_data;

  char oi[MAX_OI_LENGTH];
  char ss[MAX_SS_LENGTH];
} __attribute__((packed));
```

Is this reasonable to you? Thanks.

Reply:
Make sense to me as step 1 to be a correct work-around.

The step 2 mentioned above should be able to address the undesired overhead.

Reply:
All Done:-)